force_bdss/app/base_operation.py

```diff
@@ -0,0 +1,48 @@
+from threading import Event as ThreadingEvent
+import logging
+
+from traits.api import (
+    DelegatesTo,
+    HasStrictTraits,
+    Instance,
+    provides
+)
+
+from .i_operation import IOperation
+from .workflow_file import WorkflowFile
+
+log = logging.getLogger(__name__)
+
+
+@provides(IOperation)
+class BaseOperation(HasStrictTraits):
+    """ Base class for EvaluateOperation and OptimizeOperation.
+    """
+
+    #: The workflow file being operated on.
+    workflow_file = Instance(WorkflowFile)
+
+    #: The workflow instance.
+    workflow = DelegatesTo("workflow_file")
+
+    #: Threading Event instance that indicates if the optimization operation
+    #: should be stopped.
+    _stop_event = Instance(ThreadingEvent, visible=False, transient=True)
+
+    #: Threading Event instance that indicates if the optimization operation
+    #: should be paused and then resumed.
+    _pause_event = Instance(ThreadingEvent, visible=False, transient=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._stop_event = ThreadingEvent()
+        self._pause_event = ThreadingEvent()
+        self._pause_event.set()
+
+    def run(self):
+        """ Evaluate the workflow. """
+
+    def _deliver_event(self, event):
```

Comment:
We can include the implementation of ``_deliver_event`` in both ``EvaluateOperation`` and ``OptimizeOperation`` here, since it will be shared between both classes.

By doing so it probably allows us to use the existing unit test code for this method to meet the Codecov threshold

Reply:
Optimizeoperation._deliver_event contains a lot more code, with the code that it shares with EvaluateOperation._deliver_event coming at the end. I didn't want to screw up the order of run, etc.

Reply:
Ah, apologies - looking at this issue again I believe that the `listener` related code should also be ported to the base class. Only the `run` methods (and any related code) will end up being unique to both subclasses. Sorry, that wasn't made clear in the original issue